File: pe-gui/src/main/java/org/saar/gui/component/UIButton.java
package org.saar.gui.component;

import org.jproperty.type.BooleanProperty;
import org.jproperty.type.SimpleBooleanProperty;
import org.saar.gui.UIBlock;
import org.saar.gui.UIComponent;
import org.saar.gui.event.EventHandler;
import org.saar.gui.event.MouseEvent;
import org.saar.gui.style.Colours;

public class UIButton extends UIComponent {

    private final BooleanProperty pressedProperty = new SimpleBooleanProperty();

    private final UIBlock uiBlock = new UIBlock();

    private EventHandler<MouseEvent> onAction = e -> {};

    public UIButton() {
        initUiObject();
    }

    private void initUiObject() {
        this.uiBlock.getStyle().getBackgroundColour().set(Colours.GREY);
        this.uiBlock.getStyle().getBorderColour().set(Colours.DARK_GREY);
        this.uiBlock.getStyle().getBorders().set(2);
        add(this.uiBlock);
    }

    public BooleanProperty pressedProperty() {
        return this.pressedProperty;
    }

    public void setOnAction(EventHandler<MouseEvent> onAction) {
        this.onAction = onAction;
    }

    @Override
    public void onMousePress(MouseEvent event) {
        if (event.getButton().isPrimary()) {
            pressedProperty().set(true);
            getStyle().getColourModifier().set(1.5f);
        }
    }

    @Override
    public void onMouseRelease(MouseEvent event) {
        if (event.getButton().isPrimary()) {
            if (pressedProperty().get()) {
                pressedProperty().set(false);
                this.onAction.handle(event);
            }
            getStyle().getColourModifier().set(1);
        }
    }

    @Override
    public void onMouseEnter(MouseEvent event) {
    }

    @Override
    public void onMouseExit(MouseEvent event) {
        pressedProperty().set(false);
        getStyle().getColourModifier().set(1);
    }
}

File: pe-gui/src/main/java/org/saar/gui/component/UIButton.kt
package org.saar.gui.component

import org.jproperty.type.BooleanProperty
import org.jproperty.type.SimpleBooleanProperty
import org.saar.gui.UIComponent
import org.saar.gui.UITextElement
import org.saar.gui.event.EventHandler
import org.saar.gui.event.MouseEvent
import org.saar.gui.style.Colours
import org.saar.gui.style.value.FontSizeValues

class UIButton : UIComponent() {

    private val pressedProperty: BooleanProperty = SimpleBooleanProperty()

    private var onAction: EventHandler<MouseEvent>? = null

    override val children = listOf(
        UITextElement("Button").also { it.parent = this }.apply {
            style.fontSize.value = FontSizeValues.percent(48 / 100f)
        }
    )

    fun setOnAction(onAction: EventHandler<MouseEvent>) {
        this.onAction = onAction
    }

    override fun onMousePress(event: MouseEvent) {
        if (event.button.isPrimary) {
            this.pressedProperty.set(true)
            this.style.colourModifier.set(1.5f)
        }
    }

    override fun onMouseRelease(event: MouseEvent) {
        if (event.button.isPrimary) {
            if (this.pressedProperty.get()) {
                this.pressedProperty.set(false)
                this.onAction?.handle(event)
            }
            this.style.colourModifier.set(1f)
        }
    }

    override fun onMouseEnter(event: MouseEvent) = Unit

    override fun onMouseExit(event: MouseEvent) {
        this.pressedProperty.set(false)
        this.style.colourModifier.set(1f)
    }

    init {
        this.style.backgroundColour.set(Colours.GREY)
        this.style.borderColour.set(Colours.DARK_GREY)
        this.style.borders.set(2)
    }
}